resources/test-runner.mjs

```diff
@@ -77,3 +101,74 @@ export class TestRunner {
         return invoker.start();
     }
 }
+
+export class AsyncTestRunner extends TestRunner {
+    async runTest() {
+        // Prepare all mark labels outside the measuring loop.
+        const suiteName = this.suite.name;
+        const testName = this.test.name;
+        const syncStartLabel = `${suiteName}.${testName}-start`;
+        const syncEndLabel = `${suiteName}.${testName}-sync-end`;
+        const asyncStartLabel = `${suiteName}.${testName}-async-start`;
+        const asyncEndLabel = `${suiteName}.${testName}-async-end`;
+
+        let syncTime;
+        let asyncStartTime;
+        let asyncTime;
+
+        const runSync = async () => {
+            if (this.params.warmupBeforeSync) {
+                performance.mark("warmup-start");
+                const startTime = performance.now();
+                // Infinite loop for the specified ms.
+                while (performance.now() - startTime < this.params.warmupBeforeSync)
+                    continue;
+                performance.mark("warmup-end");
+            }
+            performance.mark(syncStartLabel);
+            const syncStartTime = performance.now();
+            await this.test.run(this.page);
```

Comment:
This is a lot of code duplication with seemingly the only difference being this line here.
I don't think we should duplicate this much code between the two runners.